src/pybind/mgr/dashboard/grafana.py

```diff
@@ -0,0 +1,113 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import json
+import os
+import time
+import requests
+
+from . import logger
+from .exceptions import GrafanaError
+from .settings import Settings
+
+
+class GrafanaRestClient(object):
+
+    def url_validation(self, method, path):
+        response = requests.request(
+            method,
+            path)
+
+        return response.status_code
+
+    def push_dashboard(self, dashboard_obj):
+        if not Settings.GRAFANA_API_URL:
+            raise GrafanaError("The Grafana API URL is not set!")
+        if not Settings.GRAFANA_API_URL.startswith('http'):
+            raise GrafanaError("The Grafana API URL is invalid")
+        if not Settings.GRAFANA_API_USERNAME:
+            raise GrafanaError("The Grafana API username is not set!")
+        if not Settings.GRAFANA_API_PASSWORD:
+            raise GrafanaError("The Grafana API password is not set!")
+        url = Settings.GRAFANA_API_URL.rstrip('/') + \
+            '/api/dashboards/db'
+        headers = {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+        }
+        payload = {
+            'dashboard': dashboard_obj,
+            'overwrite': True,
+        }
+        try:
+            response = requests.post(
+                url,
+                headers=headers,
+                data=json.dumps(payload),
+                auth=(Settings.GRAFANA_API_USERNAME,
+                      Settings.GRAFANA_API_PASSWORD),
+            )
+        except requests.ConnectionError:
+            raise GrafanaError("Could not connect to Grafana server!")
+        response.raise_for_status()
+        return response.status_code, response.json()
+
+
+class Retrier(object):
```

Comment:
That's actually the first approach I tried, but I realized it would probably not be desirable for the API endpoint or the command to potentially block for several minutes; in other words, it's really just the on-startup task that wants the retries. This way, the retry parameters are simply exposed as args that the task invocation can use.